adding ERC721 detection to TokensController

adonesky1 opened the pull request:
Until we support collectibles in the extension, we need the controller that tracks token state to be aware of whether tokens are ERC721, in order to support behavior in the extension that prevents users from attempting to send ERC721 tokens and losing gas in a transaction we cannot execute. 

This PR adds the isERC721 flag to tokens in the TokensController with all of the required methods and wiring to land it in the extension (an upcoming PR in the extension repo).

adonesky1 commented on package.json:
```diff
@@ -58,6 +58,7 @@
     "eth-sig-util": "^3.0.0",
     "ethereumjs-util": "^7.0.10",
     "ethereumjs-wallet": "^1.0.1",
+    "ethers": "^5.4.1",
```
I elected to introduce ethers to this repo because I figure we want to migrate away from web3 as our ethereum querying package.

Gudahtt replied:
We do, but we don't yet have a provider that works with ethers. It expects the provider to be EIP-1192 compliant, and the one we use internally in the extension isn't compliant with any spec :sweat_smile: It only has one method, `sendAsync`.

It's not clear exactly what works and what doesn't when it's given a provider like the one we use. The only thing I know doesn't work is that the provider breaks when the network changes. We had to destroy it and reconstruct it upon network change last time we introduced it in swaps.

adonesky1 replied:
Ok hmmm. So we were using this same passed in provider with ethers in the extension-based version of this functionality. Its initialized [here](https://github.com/MetaMask/metamask-extension/blob/a294f02b1d46a16cc443b627f2b1e492809177b1/app/scripts/controllers/preferences.js#L83). From what I can tell it works in the same way. Though, as you point out I do need to recreate a new instance on network change. Does that address concerns? Given that it mimics existing behavior?

Gudahtt replied:
Perhaps so. I'm not even sure our usage of ethers in the extension is correct though, because we don't know what other 1193 events it depends upon. Just that there are no known issues.

I can dig in to try to better understand that I suppose.

adonesky1 replied:
@Gudahtt is it okay to make that enhancement in a separate PR? Since this is not regressing the existing functionality (however limited it may be)?

Gudahtt replied:
Sorry I didn't mean to suggest any enhancements. I was talking about the usage you're introducing here in this PR.

adonesky1 replied:
Ah yes so I guess I'm thinking of "the usage [i'm] introducing here" as replicating the existing usage in the preferences controller in the extension, making the work on getting together an EIP-1193 compliant provider an enhancement in my mind. But I guess your thinking of it differently because this is a new repo where I'm introducing a wrong usage?

Gudahtt replied:
Oh, sheesh, my bad! I didn't notice that the exact same call was being made in the preferences controller already. You're absolutely right, this makes sense.

adonesky1 commented on src/assets/TokensController.ts:
```diff
@@ -178,6 +187,10 @@ export class TokensController extends BaseController<
   }
 
+  configureProvider(provider: any) {
+    this.ethersProvider = new ethers.providers.Web3Provider(provider);
```
@Gudahtt I don't know if this is the pattern we're looking for, but I saw that in other controllers you were pushing away from setters/getters for configuration effects.

Gudahtt replied:
Not sure I follow what this method is for. Isn't the provider passed in as config?

adonesky1 replied:
Yes. It should probably be and then per the comments above, reinstantiated on network change?

adonesky1 commented:
@Gudahtt sorry this PR has gotten quite large... Let me know if you want to look through it together?

Gudahtt commented on src/assets/TokensController.ts:
```diff
@@ -268,6 +295,68 @@ export class TokensController extends BaseController<
+    });
+    tokens[tokenIndex].isERC721 = await this._detectIsERC721(tokenAddress);
+    this.update({ tokens });
+    return Promise.resolve(tokens[tokenIndex]);
```
Nit: This Promise wrapper is redundant; the result will automatically be a Promise because this function is `async`.
```suggestion
    return Promise.resolve(tokens[tokenIndex]);
```

Gudahtt commented on src/assets/TokensController.ts:
```diff
@@ -268,6 +295,68 @@ export class TokensController extends BaseController<
+    return await tokenContract
+      .supportsInterface(ERC721_INTERFACE_ID)
+      .catch((error: Error) => {
+        console.log('error', error);
```
This doesn't seem great :thinking: Is this expected to throw if the interface isn't supported? What is the error in that case?

Ideally we'd be able to identify the `false` case while still throwing any unexpected errors. We don't want a network error to mistakenly identify a collectible as a token.

adonesky1 replied:
Good point. This is also a case where this is how its currently done in the extension, but I actually made this mess so I should probably clean it up, plus its a more straightforward fix than the EIP-1193 compliant provider.

adonesky1 replied:
So unfortunately the error we "expect" here is coded "UNPREDICTABLE_GAS_LIMIT". I suspect this is because these contracts don't have the supportsInterface method implemented?
![Screen Shot 2021-07-15 at 4 11 36 PM](https://user-images.githubusercontent.com/34557516/125858409-ccbccdce-6ec0-4b5a-92f0-c86dd4deea08.png)

adonesky1 replied:
Does it feel ok to return false when error.code is "UNPREDICTABLE_GAS_LIMIT"? or should we try to find a different way to check for the interface?

Gudahtt replied:
... I guess :thinking: I'm somewhat confused about how we got a gas estimation error, given that I wouldn't expect this to estimate gas.

adonesky1 replied:
Yeah... it is strange. I do see it consistently when adding non ERC721 tokens

Gudahtt commented:
It doesn't look too large, though I haven't yet looked at the tests. Thanks for the offer - I'll let message you if anything seems unclear.

Gudahtt commented on src/assets/TokenBalancesController.test.ts:
```diff
@@ -182,21 +182,28 @@ describe('TokenBalancesController', () => {
+
+    const supportsInterfaceStub = stub().returns(Promise.resolve(false));
+
+    stub(tokensController, '_createEthersContract').callsFake(() =>
```
This isn't ideal :thinking: Now the tests are tied into controller internals.

Like the APIs, I had hoped we would extract network interactions into separate modules so they could be more easily mocked in tests without needing to understand how the controller worked internally. We don't have any examples of this yet though. I don't know how to handle the provider, since unlike the APIs it is stateful.

adonesky1 replied:
Yeah I see what you mean. I'm not sure how to handle the provider here either. Maybe we can put our heads together on a call to think this through?

Gudahtt replied:
Sure! Not a blocker for this PR, but we'll want to figure out a pattern for how to do this eventually.

adonesky1 replied:
How about I document this in a separate ticket as part of a broader refactor to create a pattern for separating network interactions into modules that can be more cleanly mocked/stubbed?

Gudahtt replied:
Sure! Or rather, just add it to this list: https://github.com/MetaMask/controllers/issues/430 There is already a ticket for it.

adonesky1 replied:
Done!

Gudahtt commented on src/assets/TokensController.test.ts:
```diff
@@ -19,10 +21,21 @@ describe('TokensController', () => {
       onNetworkStateChange: (listener) => network.subscribe(listener),
     });
+
+    clock = sinon.useFakeTimers(1);
```
What is the clock used for?

adonesky1 replied:
mocking date.now() for this test: https://github.com/MetaMask/controllers/blob/74e62f1c958e1b87435a6339bd97ade8b28ede90/src/assets/TokensController.test.ts#L501

Gudahtt replied:
Thanks, that makes sense. It would be helpful if there was a comment on that line referencing the clock though, or even just that it relies upon fake timers. I had looked for a usage of that variable and found none, and thought it was unused. Timing stuff can be pretty subtle.

adonesky1 replied:
Done!

adonesky1 replied:
And actually just moved the fakeTimers call into the one test that it uses for improved visibility/clarity.

Gudahtt replied:
Nice! Thanks, this is much better.

Gudahtt commented on src/assets/TokensController.ts:
```diff
@@ -268,6 +295,68 @@ export class TokensController extends BaseController<
+    const tokenIndex = tokens.findIndex((token) => {
+      return token.address === tokenAddress;
+    });
+    tokens[tokenIndex].isERC721 = await this._detectIsERC721(tokenAddress);
```
The token may have been deleted by the time this call resolves. Or it might be at a different index.

It would be safer to do this in a different order. Maybe by calling this first and storing the result in a local variable, then finding the token and setting the property if it's found.

Or we could use a lock, but it'd be faster to avoid locks if possible.

adonesky1 replied:
addressed!

Gudahtt commented on src/assets/TokensController.ts:
```diff
@@ -268,6 +295,68 @@ export class TokensController extends BaseController<
+   * Adds isERC721 field to token object
+   * (Called when a user attempts to add tokens that were previously added which do not yet had isERC721 field)
+   *
+   * @param {string} tokenAddress - The contract address of the token requiring the isERC721 field added.
+   * @returns {Promise<object>} The new token object with the added isERC721 field.
```
Nit: Usually in TypeScript repos we omit the type, because it's already documented in the type signature.

```suggestion
   * @param tokenAddress - The contract address of the token requiring the isERC721 field added.
   * @returns The new token object with the added isERC721 field.
```

adonesky1 replied:
addressed!

Gudahtt commented on src/assets/TokensController.ts:
```diff
@@ -268,6 +295,68 @@ export class TokensController extends BaseController<
+   * Detects whether or not a token is ERC-721 compatible.
+   *
+   * @param {string} tokensAddress - the token contract address.
+   *
```
Nit: The return type is missing from this doc comment

adonesky1 replied:
addressed

Gudahtt commented on src/assets/TokensController.ts:
```diff
@@ -268,6 +294,72 @@ export class TokensController extends BaseController<
+      this.ethersProvider,
+    );
+
+    return await tokenContract
+      .supportsInterface(ERC721_INTERFACE_ID)
+      .catch((error: { code: string; method: string }) => {
+        if (error.code === 'UNPREDICTABLE_GAS_LIMIT') {
+          return false;
+        }
+        throw error;
+      });
```
Nit: Using `await` rather than chaining Promises has better support for async stack traces in Chrome (and hopefully soon in Firefox).

```suggestion
    try {
      return await tokenContract
        .supportsInterface(ERC721_INTERFACE_ID);
    } catch (error: Error) {
        if (error.code === 'UNPREDICTABLE_GAS_LIMIT') {
          return false;
        }
        throw error;
    }
```

Also the error type should be `Error`. It has more than just those two properties, and various libraries like Sentry to treat it differently.

Gudahtt replied:
Also, we should have a comment explaining that we don't know why `UNPREDICTABLE_GAS_LIMIT` is thrown here, but that it's what we see for ERC20 tokens. That could be confusing later!

adonesky1 replied:
Ok. I originally had the type set to `Error` but it doesn't support code as a field?

adonesky1 replied:
Actually it looks like we can't specify an error type in a try/catch in ts? That seems wrong...
![Screen Shot 2021-07-16 at 9 32 12 AM](https://user-images.githubusercontent.com/34557516/125964079-ec228630-5555-4f57-9083-a5760cfd8730.png)

Gudahtt replied:
Huh, TIL. I'm now noticing that I haven't actually declared a type on a catch block before, it's always been inferred.

`unknown` probably would be best because you can throw anything really. Just that it's very inconvenient for debugging if anything but an `Error` is thrown. So I guess I was hoping declaring it as an `Error` would make more likely that TypeScript would attempt to make it so, but that's not how any of this works.

adonesky1 replied:
Think it may have to be `any` :(
![Screen Shot 2021-07-16 at 9 58 05 AM](https://user-images.githubusercontent.com/34557516/125967773-f1cee71e-b9bd-47c3-8e7a-b49b4903c100.png)

adonesky1 replied:
Made type `any` for now...

Gudahtt replied:
That error is the TypeScript compiler asking you to further narrow the type. Something like `typeof error === 'object' &&` might do the trick! Though to be honest I would have expected the optional chain operator to work :thinking: Strange.

Gudahtt replied:
Wow, I gave this a shot and it's really not making this easy :grimacing:  Not blocking on this, but I did think the whole point of the `unknown` type was to force you to not make assumptions about the type, and... I think we're doing that well here despite the compiler's complaints. So that's a shame.

Gudahtt replied:
I made a thread about this in the dev channel. I think I found a solution but we can pursue this some other time.

Gudahtt commented on src/assets/TokensController.test.ts:
```diff
@@ -35,19 +48,22 @@ describe('TokensController', () => {
   });
 
   it('should add token', async () => {
+    supportsInterfaceStub = sinon.stub().returns(Promise.resolve(false));
```
What is this trying to do :thinking: 

It seems like you're trying to make the `supportsInterface` return `false` when it's called in this test, but I don't see how this could work. `supportsInterfaceStub` was `undefined` when `beforeEach` was run, and reassigning the variable here won't change what was passed into the `callsFake` call. I suspect this is only affecting the _next_ test, not this one.

But that doesn't make sense either - it seems to work later on :thinking: I am confused.

Either way, shared mocks like this should not be used in unit tests. It introduces a subtle dependency between tests, and it can make them very difficult to understand.

Gudahtt replied:
Oh I think I understand. The `callsFake` function isn't executed until later, after it has been reassigned. 

Still seems bad for tests to depend upon code run in previous tests though.

adonesky1 replied:
Yeah makes sense. Was trying to save some duplicated code but I will add the full stub to each test that requires it.

Gudahtt commented on src/assets/TokensController.test.ts:
```diff
@@ -34,24 +37,38 @@ describe('TokensController', () => {
-  it('should add token', async () => {
+  it('should add a token', async () => {
+    const supportsInterfaceStub = sinon.stub().returns(Promise.resolve(false));
+    await sinon
```
I don't think `await` is needed here :thinking: The fake function returns a Promise, but it's just being put in place here, not executed.

adonesky1 replied:
You're right! Not sure why I thought that was necessary!

adonesky1 replied:
removed throughout the test

Gudahtt commented on src/assets/TokensController.test.ts:
```diff
@@ -312,4 +293,458 @@ describe('TokensController', () => {
+        );
+    });
+
+    afterEach(function () {
+      sinon.restore();
+    });
+
```
This seems redundant, since `sinon.restore()` is called in the `beforeEach` in the top-level `describe` block

```suggestion
```

Gudahtt commented on src/assets/TokensController.test.ts:
```diff
@@ -312,4 +293,458 @@ describe('TokensController', () => {
+        },
+      ]);
+    });
+    it('should add a valid suggested asset via watchAsset', async () => {
```
This seems like a duplicate of `should handle ERC20 type and add to suggestedAssets` :thinking:

adonesky1 replied:
Good call! Removing

Gudahtt reviewed (commented):
Found a couple of things to clean up in the unit tests,but everything else looks good!

adonesky1 commented on src/assets/TokensController.ts:
```diff
@@ -326,7 +326,10 @@ export class TokensController extends BaseController<
       return Promise.resolve(true);
+    } else if (contractsMap[checksumAddress]?.erc20 === true) {
+      return Promise.resolve(false);
     }
```
@Gudahtt  I've add this block to avoid unnecessary network requests in this method when we know the token is ERC20 because its in our contract map.

Gudahtt replied:
Good idea!

adonesky1 commented:
@Gudahtt thanks again for your always thorough reviews! I've cleaned up the tests, made them more thorough for all possible scenarios around the isERC721 flag functionality and made them only point to public functions. Also added a conditional block in the _detectIsERC721 to do the same check against our contract-maps for ERC20 tokens as we do for ERC721 to avoid unnecessary network requests for data we already have.

Gudahtt reviewed (approved):
LGTM!
